Implement mountpoints cache mechanism

cofyc opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This implements a mountpoints cache mechanism on linux platform to improve mountpoints check performance. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #49793 

**Special notes for your reviewer**:

Mounter.List() will use cache if cache is fresh. When Mounter calls Mount or Unmount and it succeeds, it invalidates cache immediately. Then in next List() call, mounter will try to get new mountpoints and refresh the cache again.

Though each node there is only one kubelet process running, administrators or third party utilities may mount/unmount mountpoints, so we should watch changes of mount tree. If mount tree changes, we need to invalidate the cache. At kubelet startup, a loop is started to poll on "/proc/mounts" file. If it receives a change event, it invalidates the mountpoints cache.

Polling on `/proc/mounts` file to get notified when mount tree change is a reliable way to go. See `proc` manpage about "/proc/[pid]/mounts" http://man7.org/linux/man-pages/man5/proc.5.html and related linux commits:

- https://github.com/torvalds/linux/commit/5addc5dd8836aa061f6efc4a0d9ba6323726297a
- https://github.com/torvalds/linux/commit/31b07093c44a7a442394d44423e21d783f5523b8
- https://github.com/torvalds/linux/commit/033b96fd30db52a710d97b06f87d16fc59fee0f1

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
Implement mountpoints cache mechanism on linux platform to improve
mountpoints check performance.

Mounter.List() will use cache if cache is fresh. At kubelet startup, a
loop is started to poll on "/proc/mounts" file. If it receives a change
event, it invalidates the mountpoints cache. Then in next List() call,
mounter will try to get new mountpoints and refresh the cache again.

cofyc commented:
/assign @msau42

Commit:
Use a dedicate interface for mounter with mountpoints cache.

gnufied commented:
@cofyc awesome stuff. Do you have some benchmarks of before and after number on a busy node? It would be interesting to know how much it affects pod startup performance in general on a busy node.

msau42 commented:
/ok-to-test

One more thing to consider especially for local volumes.  Bind-mounts are not detected in /proc/mounts.  You need to parse /proc/self/mountinfo to see bind mounts.

msau42 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1384,6 +1384,11 @@ func (kl *Kubelet) Run(updates <-chan kubetypes.PodUpdate) {
 	}
 
+	// Start mounter if it implements mount.MounterWithMountPointsCache.
+	if mounterCache, ok := kl.mounter.(mount.MounterWithMountPointsCache); ok {
```
Can you put this behind some configuration flag so we have the option to turn it off if we find an issue with this in production?

cofyc replied:
ok, I'll add it.

cofyc replied:
@msau42 Added in https://github.com/kubernetes/kubernetes/pull/61776/commits/7a245922ec46020261de1b8aac4c795de5cc426d.

Commit:
Fix a few bugs in mountpoints_cache_linux.go.

- Double-checking locking criterion.
- Close fd on exit.
- Add default case for select.

msau42 commented:
Agree with @gnufied.  Would like to see some benchmark numbers to show that this will actually improve performance.

cofyc commented:
@gnufied @msau42 Didn't benchmark it yet. I'll do it soon.

cofyc commented:
/test pull-kubernetes-bazel-build

Commit:
Add a flag `--disable-mountpoints-cache` to disable mountpoints cache
mechanism.

cofyc commented:
@gnufied @msau42 

hi, I've done some benchmarks and profiling. 

My testing environment:

```
1 master and 3 workers.
OS: Ubuntu 16.04
Kernel: 4.4.0-112-generic
CPU: 8
Memory: 32G (master) 16G (worker)
kubelet versions: 
  - kubelet.v1.11.0-alpha.0.1878+acd435c57853ad
  - kubelet.v1.11.0-alpha.0.1882+7a245922ec4602 (with mountpoints cache patch)
```

I use [clusterloader](https://github.com/kubernetes/perf-tests/tree/master/clusterloader) to create 300 pods sequentially. Each pod mounts 3 local volumes. (clusterloader config: https://pastebin.com/raw/RAPMMPUe and pod template: https://pastebin.com/raw/AxAnu3wW).

## Storage Operation Duration and Pod Start Latency

During pods are creating, I use prometheus to scrape kubelet metrics and grafana to visualize them.

Here is 99th percentile storage operation duration of local-volume plugin.

![storage-operation-duration](https://user-images.githubusercontent.com/91730/38194131-3858bafa-36a8-11e8-9c17-dcb98a417b40.png)

Grafana query: `histogram_quantile($percentile, sum(rate(storage_operation_duration_seconds_bucket[2m])) by (le, volume_plugin, operation_name))`.

Here is pod start latency:

![pod-start-latency](https://user-images.githubusercontent.com/91730/38194165-7581d25e-36a8-11e8-986b-ca0e8151fad1.png)

Gafana query: `rate(kubelet_pod_start_latency_microseconds_sum[2m]) / rate(kubelet_pod_start_latency_microseconds_count[2m])`

## Go profiling

I also used go pprof tool to profile kubelet: (filtered by `-nodefraction=0 -edgefraction=0 -focus \(\*Mounter\).List`)  

kubelet-before:

![cpu1](https://user-images.githubusercontent.com/91730/38192078-dca41754-369d-11e8-9c52-806b0cdc0cc0.png)

kubelet-after-patching:
![cpu2](https://user-images.githubusercontent.com/91730/38192230-89210c76-369e-11e8-913d-f041d5506982.png)

Before patching, `(*localVolumeMounter).SetUpAt` consumes 1.35s (53.15%) in `(*Mounter).IsNotMounPoint` and 1.19s (46.85%) in `GetMountRefsbyDev`.

After patching, `(*localVolumeMounter).SetUpAt` consumes 1.12s (99.12%) in `(*Mounter).IsNotMounPoint` and 0.01s (0.88%) in `GetMountRefsbyDev`.

With mountpoints cache, it reduces the duplicated `(*Mounter).List` calls if mountpoints does not change. But from graphs, I didn't see much performance improvement.

gnufied commented on pkg/util/mount/mountpoints_cache_linux.go:
```diff
@@ -0,0 +1,137 @@
+	c.Lock()
+	defer c.Unlock()
+	// Double-checking locking criterion.
+	if !c.stale {
```
Why do we need this? Once this function enters this section - the caller goroutine will have exclusive lock on code that could remove stale flag right?

cofyc replied:
The first lock criterion is checked with reader lock not a exclusive lock. It's optimistic lock hint. We need to check the locking criterion again before proceeding the actual locking logic.

e.g. It's possible A and B acquired reader lock and found `c.stale` is true at the same time. Both of them will try to lock and proceed. If we don't checking locking criterion again, `list()` will be processed twice.

gnufied replied:
Cool - I see that you just wanted to return early in second goroutine if first goroutine acquired lock and updated the cache.

gnufied commented on pkg/util/mount/mountpoints_cache_linux.go:
```diff
@@ -0,0 +1,137 @@
+			// timed out, poll again
+			continue
+		}
+		if pollfds[0].Revents&unix.POLLERR != 0 {
```
What happens if rather than calling `markStale` here - we also update the mount point cache? In this way - mount point cache is pretty much always fresh except perhaps the small time duration in which we read the `/proc/mounts` and it gets updated.

cofyc replied:
When it tries to keep cache up to date here, user of `Mounter.List()` can use cache directly. But I don't know this will be better or not, because`/proc/mounts` may be processed unnecessarily. e.g. "r w w w r" (w=mount/unmount, r=listMounts), each `w` will trigger it to process "/proc/mounts" again.

At present, I prefer to keep it simple.

msau42 commented:
@cofyc this is cool, but I agree, if the numbers are not showing much improvement, I'm not sure about the benefit of adding it.  I think the work you did to profile and analyze is really useful to everyone though. Let's find a place where you could document the steps.

cofyc commented:
@msau42 

> One more thing to consider especially for local volumes. Bind-mounts are not detected in 
> /proc/mounts. You need to parse /proc/self/mountinfo to see bind mounts.
> Edit: Nevermind it looks like mounter.List() does return bind mounts.

This is partially true, bind-mounts are detected in `/proc/mounts`, but it does not contain root of mount for bind mounts. So `mount.GetMountRefsByDev()` will not work in some cases. e.g.

```
# mkdir /tmp/src /mnt/dst
# mount --bind /tmp/src/ /mnt/dst
# grep -P 'src|dst' /proc/mounts 
/dev/mapper/vagrant--vg-root /mnt/dst ext4 rw,relatime,errors=remount-ro,data=ordered 0 0
# grep -P 'src|dst' /proc/self/mountinfo 
226 24 253:0 /tmp/src /mnt/dst rw,relatime shared:1 - ext4 /dev/mapper/vagrant--vg-root rw,errors=remount-ro,data=ordered
```

`mount.GetMountRefsByDev("/tmp/src")` cannot get bind mounts of "/tmp/src". I will fix this in pr for https://github.com/kubernetes/kubernetes/issues/61741.

cofyc commented:
Due to not showing much improvements, close it.
